### PR TITLE
Link to a path

### PR DESCRIPTION
**Pull request by acanimal**
In addition to link to a route with params we can also link to a given path that conforms a route pattern.

In my case I have complex route patterns which extracts some variables from the path. I want to be able to Link to a different path and use the power of regexp to extract again the variables from the path instead to compute the route with params.

### CODE REVIEW

**Comment by coluccini** [comment]
:+1:
This would be really useful when working with API that returns full paths

**Comment by acanimal** [comment]
Hi @fridays, this PR is important for me. I would appreciate if you could review and, if accepted, publish a new npm package version.

**Comment by HaNdTriX** [comment]
Hey @acanimal 

Just a short question what exactly is the difference between:

```javascript
Router.push('/somePath?query=123')
```

```javascript
Router.pushPath('/somePath?query=123')
```

**Comment by acanimal** [comment]
@HaNdTriX `Router.next()` will try to find a next page with the given path. In contrast with `Router.pushPath` we can pass a path and found the router that match that regexp.

In our case we have different routes that must render the same next page. That methods helps us on that tasks.

**Comment by coluccini** [comment]
For example:
```javascript
routes.add('someRouteName', '/:path', 'otherPage');

Router.push('/somePath?query=123')
// -> Search por /pages/somePath
Router.pushPath('/somePath?query=123')
// -> Search por /pages/otherPage
```

**Comment by fridays** [comment]
Thank you for contributing!

It's great and we should have this feature.

What do you all think about naming it `href` instead of `path` as suggested in https://github.com/fridays/next-routes/pull/23? Then if no matching route was found, we could assign `newProps.href` and next.js would render it’s error page.

File: README.md [hunk 1]
```diff
@@ -91,10 +91,26 @@ export default () => (
...
+  </div>
+)
+
+```
+
 API: `<Link route="name" params={params}>...</Link>`
 
 - `route` - Name of a route
 - `params` - Optional parameters for the route URL
+- `path`- A path supposed to follow a route pattern. From the path the right route and params are found. NOTE: `path`and `route` exclusive and if present `route` takes precedence over `path`.
```
Comment (fridays) [inline README.md]: Please clarify the last part/intention, the implementation seems that `path` takes precedence.

File: README.md [hunk 2]
```diff
@@ -127,6 +143,8 @@ API:
 
 `Router.pushRoute(name, params, options)`
 
+`Router.pushRoute(path, options)`
```
Comment (fridays) [inline README.md]: Small typo here `pushRoute`

File: src/index.js [hunk 3]
```diff
@@ -55,18 +55,29 @@ class Routes {
...
   getLink (Link) {
     const LinkRoutes = props => {
-      const {route, params, ...newProps} = props
+      const {route, params, path, ...newProps} = props
 
       if (route) {
         Object.assign(newProps, this.findByName(route).getLinkProps(params))
       }
 
+      if (path) {
```
Comment (fridays) [inline src/index.js]: Should this be `else if`?

File: src/index.js [hunk 4]
```diff
@@ -55,18 +55,29 @@ class Routes {
...
+        const {route: routeObj, params: paramsObj} = this.match(path)
```
Comment (fridays) [inline src/index.js]: We need to check `if (routeObj)`. If no route was found, maybe assign `newProps.href`

File: src/index.js [hunk 5]
```diff
@@ -55,18 +55,29 @@ class Routes {
...
+      }
+
       return <Link {...newProps} />
     }
     return LinkRoutes
   }
 
   getRouter (Router) {
+    Router.pushPath = (path, options) => {
+      const {route: routeObj, params: paramsObj} = this.match(path)
```
Comment (fridays) [inline src/index.js]: Needs `if (routeObj)` here too

File: README.md [hunk 6]
```diff
@@ -91,10 +91,26 @@ export default () => (
...
 ```
 
+```jsx
+// pages/index.js
+import {Link} from '../routes'
+
+export default () => (
+  <div>
+    <div>Welcome to next.js!</div>
+    <Link path='/some-path-conform-to-a-route-patter'>
```
Comment (fridays) [inline README.md]: Please just add the `Link` into the upper example

**Comment by acanimal** [comment]
I agree with the change from path to href. I'll try to made all changes ASAP and update the PR. Thanks.

**Comment by acanimal** [comment]
@fridays Changes done !!!

**Comment by acanimal** [comment]
@HaNdTriX @fridays any news?

**Comment by acanimal** [comment]
@HaNdTriX @fridays any news?

**Comment by fridays** [comment]
Here it's calling `this.match(href)` but it may be undefined:
https://github.com/acanimal/next-routes/blob/a1b8defb43f29b515521f908832501999962f603/src/index.js#L59

I'm not sure if `href` should take precedence over `route`, could you share the reasoning?

Another catch: It doesn't yet take the querystring from a given `href` into account. It has to be parsed and merged with params like in the request handler for consistent results.

**Comment by acanimal** [comment]
@fridays yes, `this.match(href)` can be undefined because of this I check on https://github.com/acanimal/next-routes/blob/a1b8defb43f29b515521f908832501999962f603/src/index.js#L64 if match returns some router object then I apply them otherwise use the router property.

There is no special reason why I decide href takes precedence over route. I simply though an href written by the user was more important than a route with params. I think is a more "direct" way to specify so it must take precedence.

Anyway it is not important. I can change if you do not agree.

Totally agree with the fact we are not parsing query params. This little change is really urgent for us. If you agree you could merge to be the changes and I can continue working with the query params.

**Comment by acanimal** [comment]
@HaNdTriX @fridays any news?

**Comment by fridays** [comment]
Hi @acanimal I mean `href` can be undefined, it should not try to match it then.

We should add this feature once complete, see my last comment regarding querystring. Also the other Router methods are not there yet. I can have a look at the missing parts.

If you're in a hurry you can of course `npm install your/repo` or use a HOC for now

**Pull request closed without merging**